### ISSUE TITLE
API design tweak for tracking generic methods

### ISSUE THREAD

**Issue by jmesserly** [dart-lang/sdk#26271]
TL;DR -- here is a proposal
- remove staticInvokeType (and propagatedInvokeType) from InvocationExpression (and subtypes)
- introduce `List<DartType> inferredTypeArguments` on InvocationExpression

---

Right now in Analyzer, given a generic function/method, MethodInvocation and FunctionExpressionInvocation will track "staticInvokeType" (also "propagatedInvokeType" but we can ignore that for now). For generic tear offs, we store it in staticType (because it's the result type for that expression).

For example given `<T>T -> T` with arguments `<int>` it will store `int -> int`. This is convenient for Analyzer itself, because it can easily read off the returnType or the parameter types, and get the correct value. (Avoiding a common class of bug where we got `T` when we should've gotten `int`.)

However, there's a bit of redundancy in that representation, and sometimes you really just want to read off the type parameters. For example, in DDC code generation, I really just want to know `<int>` (regardless of whether it was inferred or written down).

Anyway, given `<T>T -> T` and `int -> int` I can figure out that we have `int/T` pretty easily. But it'd be nice if it was stored explicitly somewhere.

CC @leafpetersen who's thought about this as well. Not urgent, I have a workaround for DDC. It's kind of hacky but so far seems to get the job done:

``` dart
  /// Given a generic function type [g] and an instantiated function type [f],
  /// find a list of type arguments TArgs such that `g<TArgs> == f`,
  /// and return TArgs.
  ///
  /// This function must be called with type [f] that was instantiated from [g].
  Iterable<DartType> _recoverTypeArguments(FunctionType g, FunctionType f) {
    // TODO(jmesserly): this design is a bit unfortunate. It would be nice if
    // resolution could simply create a synthetic type argument list.
    assert(identical(g.element, f.element));
    assert(g.typeFormals.isNotEmpty && f.typeFormals.isEmpty);
    assert(g.typeFormals.length + g.typeArguments.length ==
        f.typeArguments.length);

    // Instantiation in Analyzer works like this:
    // Given:
    //     {U/T} <S> T -> S
    // Where {U/T} represents the typeArguments (U) and typeParameters (T) list,
    // and <S> represents the typeFormals.
    //
    // Now instantiate([V]), and the result should be:
    //     {U/T, V/S} T -> S.
    //
    // Therefore, we can recover the typeArguments from our instantiated
    // function.
    return f.typeArguments.skip(g.typeArguments.length);
  }

```

**Comment by jmesserly** [comment]
this would be a good change, but I don't think it's high priority now.
